### PR TITLE
Slightly speed up the `resize_inner` function + documentation of other functions.

### LINKED ISSUE #453: Bug in `RawIter::reflect_toggle_full`
This method uses comparison of indexes (pointers) through the `Bucket::as_ptr` function.

This logic only works if `T` is not zero sized type

### PR DESCRIPTION
**Pull request by JustForFun88**
This speeds up the `resize_inner` function a bit, since now reading the data from the heap is done not by byte, but by a group. In addition, we may not iterate over all control bytes if we have yielded all indexes of full buckets. For example, on my computer:

Before (with `cargo bench`):
```
test grow_insert_ahash_highbits  ... bench:      32,481 ns/iter (+/- 5,315)
test grow_insert_ahash_random    ... bench:      30,122 ns/iter (+/- 5,568)
test grow_insert_ahash_serial    ... bench:      34,267 ns/iter (+/- 8,364)
test grow_insert_std_highbits    ... bench:      54,410 ns/iter (+/- 5,030)
test grow_insert_std_random      ... bench:      52,566 ns/iter (+/- 6,455)
test grow_insert_std_serial      ... bench:      51,681 ns/iter (+/- 9,595)
```

After (with `cargo bench`):
```
test grow_insert_ahash_highbits  ... bench:      27,597 ns/iter (+/- 1,436)
test grow_insert_ahash_random    ... bench:      27,096 ns/iter (+/- 203)
test grow_insert_ahash_serial    ... bench:      27,369 ns/iter (+/- 719)
test grow_insert_std_highbits    ... bench:      50,399 ns/iter (+/- 449)
test grow_insert_std_random      ... bench:      44,495 ns/iter (+/- 7,180)
test grow_insert_std_serial      ... bench:      47,735 ns/iter (+/- 7,598)
```

As for the documentation, I started with `resize_inner` and since this function depends on others, I had to document them as well, etc., so there a lot of documentation in total. Also fixed a couple of inaccuracies with marking the function as unsafe.

Fix #453.

### CODE REVIEW

File: src/raw/mod.rs [hunk 1]
```diff
@@ -2211,11 +2304,70 @@ impl<A: Allocator + Clone> RawTableInner<A> {
         }
     }
 
+    /// Returns an iterator over full buckets indices in the table.
+    ///
+    /// # Safety
+    ///
+    /// The caller to ensure that the `RawTableInner` outlives the
```
Comment (Zoxc) [inline src/raw/mod.rs:2311]: `has to ensure`?

File: src/raw/mod.rs [hunk 2]
```diff
@@ -2211,11 +2304,70 @@ impl<A: Allocator + Clone> RawTableInner<A> {
...
+    /// `FullBucketsIndices`. Because we cannot make the `next` method
+    /// unsafe on the `FullBucketsIndices` struct, we have to make the
+    /// `full_control_bytes_iter` method unsafe.
```
Comment (Zoxc) [inline src/raw/mod.rs:2314]: By `full_control_bytes_iter` did you mean this method (`full_buckets_indices`)?

File: src/raw/mod.rs [hunk 3]
```diff
@@ -2152,14 +2238,12 @@ impl<A: Allocator + Clone> RawTableInner<A> {
...
         // Allocate and initialize the new table.
-        let mut new_table = RawTableInner::fallible_with_capacity(
+        let new_table = RawTableInner::fallible_with_capacity(
             self.alloc.clone(),
             table_layout,
             capacity,
             fallibility,
         )?;
-        new_table.growth_left -= self.items;
-        new_table.items = self.items;
```
Comment (Zoxc) [inline src/raw/mod.rs:2162]: Why are these moved?

**Comment by Zoxc** [comment]
I tested this PR out with rustc and it does seem to be a performance improvement there:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.6201s</td><td align="right">1.6077s</td><td align="right"> -0.77%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2400s</td><td align="right">0.2389s</td><td align="right"> -0.47%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9139s</td><td align="right">0.9061s</td><td align="right"> -0.85%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5023s</td><td align="right">1.4907s</td><td align="right"> -0.77%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">5.6779s</td><td align="right">5.6401s</td><td align="right"> -0.67%</td></tr><tr><td>Total</td><td align="right">9.9542s</td><td align="right">9.8835s</td><td align="right"> -0.71%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">0.9930s</td><td align="right"> -0.70%</td></tr></table>

File: src/raw/mod.rs (hunk 3 above)
Comment (JustForFun88) [reply src/raw/mod.rs:2162]: They are not needed here because:
1. This allows you to make the function safe.
2. Adds more consistency in functions, allowing less overthinking and remembering that you changed the number of elements before actually adding them, which is not a good idea in my opinion and only confuses. For example, in the `clone_from_impl` method, we first add items, and only then change `self.table.items` and `self.table.growth_left` fields.

File: src/raw/mod.rs (hunk 1 above)
Comment (JustForFun88) [reply src/raw/mod.rs]: Corrected a typo, thanks

File: src/raw/mod.rs (hunk 2 above)
Comment (JustForFun88) [reply src/raw/mod.rs]: Done

File: src/raw/mod.rs [hunk 4]
```diff
@@ -2141,9 +2208,28 @@ impl<A: Allocator + Clone> RawTableInner<A> {
...
+    ///
+    /// If `table_layout` does not match the `TableLayout` that was used to allocate
+    /// this table, then using `mem::swap` with the `self` and the new table returned
+    /// by this function results in [`undefined behavior`].
+    ///
+    /// [`undefined behavior`]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
     #[allow(clippy::mut_mut)]
     #[inline]
-    unsafe fn prepare_resize(
+    fn prepare_resize(
```
Comment (Amanieu) [inline src/raw/mod.rs:2232]: It occurs to me that a lot of these methods that mutate data on the heap should probably take `&mut self`, otherwise there is an additional safety requirement that no other thread is concurrently accessing the data.

File: src/raw/mod.rs [hunk 5]
```diff
@@ -3065,6 +3306,125 @@ impl<T> Iterator for RawIter<T> {
...
+/// must observe several rules when using it:
+/// - You must not free the hash table while iterating (including via growing/shrinking).
+/// - It is fine to erase a bucket that has been yielded by the iterator.
+/// - Erasing a bucket that has not yet been yielded by the iterator may still
+///   result in the iterator yielding index of that bucket.
+/// - It is unspecified whether an element inserted after the iterator was
+///   created will be yielded by that iterator.
+/// - The order in which the iterator yields indices of the buckets is unspecified
+///   and may change in the future.
+pub(crate) struct FullBucketsIndices {
```
[the remark this answers is not in the document]
Comment (Amanieu) [reply src/raw/mod.rs:3321]: Could `RawIter` be built on top of this to avoid code duplication? Assuming that doesn't impact compilation times too much due to the additional layer of inlining needed.

[the remark this answers is not in the document]
Comment (Amanieu) [reply src/raw/mod.rs:3321]: Actually I think this would improve compile times since the code for iteration would only need to be instantiated once.

[the remark this answers is not in the document]
Comment (Zoxc) [reply src/raw/mod.rs:3321]: I was thinking that `FullBucketsIndices` could be changed to take a size field and thus maintain a pointer to the bucket. There could be some performance impact due to the use of indices that `RawIter` could be sensitive too.

I'd try these changes on top of this PR though.

[the remark this answers is not in the document]
Comment (JustForFun88) [reply src/raw/mod.rs:3321]: I will try 😄

[the remark this answers is not in the document]
Comment (JustForFun88) [reply src/raw/mod.rs:3321]: Oh, that `reflect_toggle_full` method of `RawIter`.  It causes the `FullBucketsIndices` size to bloat.  Okay, I'll go to bed, maybe tomorrow I'll come up with something adequate 🤔

**Comment by JustForFun88** [comment]
@Amanieu I implemented `RawIter` on top of `FullBucketsIndices`, could you please review the PR?
Of course, we can go further and try to built `RawIterRange` on top of `FullBucketsIndices` too 😄 (although I haven't thought about it yet).

@Zoxc Could you please test this PR with rustc again?

**Comment by Zoxc** [comment]
Testing just the `RawIter` change:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.6238s</td><td align="right">1.6322s</td><td align="right"> 0.51%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2345s</td><td align="right">0.2354s</td><td align="right"> 0.37%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9129s</td><td align="right">0.9183s</td><td align="right"> 0.59%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.4699s</td><td align="right">1.4747s</td><td align="right"> 0.32%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">5.6903s</td><td align="right">5.7217s</td><td align="right"> 0.55%</td></tr><tr><td>Total</td><td align="right">9.9315s</td><td align="right">9.9821s</td><td align="right"> 0.51%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">1.0047s</td><td align="right"> 0.47%</td></tr></table>

**Comment by JustForFun88** [comment]
> Testing just the `RawIter` change:

I think I found the cause of the regression. Can you please repeat?

**Comment by Zoxc** [comment]
Still a regression:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.6099s</td><td align="right">1.6159s</td><td align="right"> 0.37%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2346s</td><td align="right">0.2356s</td><td align="right"> 0.40%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.8979s</td><td align="right">0.9019s</td><td align="right"> 0.45%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.4505s</td><td align="right">1.4569s</td><td align="right"> 0.44%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">5.6328s</td><td align="right">5.6628s</td><td align="right"> 0.53%</td></tr><tr><td>Total</td><td align="right">9.8257s</td><td align="right">9.8730s</td><td align="right"> 0.48%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">1.0044s</td><td align="right"> 0.44%</td></tr></table>

**Comment by JustForFun88** [comment]
@Amanieu I give up 😄. As @Zoxc predicted, and then confirmed by benchmarking, the implementation of `RawIter` through `FullBucketsIndices` leads to regression. I don't think this can be changed, because in fact one operation `bucket.next_n(index)` is replaced by:
```rust
let data_index = group_first_index + index;
bucket.next_n(data_index)
```
That is, each time there is an additional `add` operation, which leads to a small but measurable regression.

To reduce the code itself, we can, of course, perform some sort of dispatching through traits and generics, like the example below, but this will likely slow down compilation and, in essence, will be equivalent to two different structures:
```rust
trait GroupBase {
    fn next_n(&self, offset: usize) -> Self;
}

impl GroupBase for usize {
    fn next_n(&self, offset: usize) -> Self {
        self + offset
    }
}

impl<T> GroupBase for Bucket<T> {
    fn next_n(&self, offset: usize) -> Self {
        unsafe { self.next_n(offset) }
    }
}

pub(crate) struct FullBucketsIndices<B: GroupBase> {
    current_group: BitMaskIter,
    group_base: B,
    // Pointer to the current group of control bytes,
    ctrl: NonNull<u8>,
    // Buckets number in given subset of a table.
    buckets: usize,
    // Number of elements in given subset of a table.
    items: usize,
}
```

File: src/raw/mod.rs (hunk 4 above)
Comment (JustForFun88) [reply src/raw/mod.rs:2242]: I can try to go through all the functions (after the PR), although it is probably won’t be able to do this with all of them.

Specifically, this function can be left as it is (with `&self`), since it does not change the original table (neither before nor after this pull request).

**Comment by Amanieu** [comment]
That's fine, let's just switch back to the previous implementation of `RawIter` then.

**Comment by JustForFun88** [comment]
> That's fine, let's just switch back to the previous implementation of `RawIter` then.

Returned the previous implementation. However, the tests are failing, it looks like an upstrem problem [rust-lang/rust#115239](https://github.com/rust-lang/rust/issues/115239)

**Comment by Amanieu** [comment]
LGTM. Let's wait until CI is resolved before merging.
